### PR TITLE
extract_alarm_telecare.R

### PR DESCRIPTION
**Pull request by lizihao-anu**
A draft for extract_alarm_telecare.R, which converts B17 spss script to R.

### CODE REVIEW

File: Production_scripts/Process_extracts/extract_alarm_telecare.R [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    smrtype,
+    sc_latest_submission
+  ) %>%
+  # end of lazy_dt()
+  as_tibble() %>%
+  # Sort for running SPSS
+  arrange(
+    sending_location,
+    social_care_id
+  )
```
Comment (Moohan) [inline Production_scripts/Process_extracts/extract_alarm_telecare.R:172]: This should be in a separate script - only the below is dependant on year

We will have 2 scripts; This code should be in `"Production_scripts/Social_Care/alarm_telecare.R"` and the year specific code will be in `"Production_scripts/Process_extracts/extract_alarm_telecare.R"`

File: Production_scripts/Process_extracts/extract_alarm_telecare.R [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+
+# Transform B17 ----
+
+# Now select epsiodes for given FY
+
+# Is there any function for fetch this address or should we write one?
+client_table_path <- fs::path(
+  glue::glue(
+    "/conf/sourcedev/Source_Linkage_File_Updates/{year}/Client_for_Source-20{year}"
```
[the remark this answers is not in the document]
Comment (Moohan) [reply Production_scripts/Process_extracts/extract_alarm_telecare.R:182]: `readr::read_rds(get_source_extract_path(year, type = "Client"))`

File: Production_scripts/Process_extracts/extract_alarm_telecare.R [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    sc_support_from_unpaid_carer,
+    sc_social_worker,
+    sc_type_of_housing,
+    sc_meals,
+    sc_day_care
+  ) %>%
+  haven::write_sav(fs::path(
+    glue::glue("/conf/sourcedev/Source_Linkage_File_Updates/{year}/Alarms-Telecare-for-source-20{year}"),
+    ext = "zsav"
+  ))
```
Comment (Moohan) [inline Production_scripts/Process_extracts/extract_alarm_telecare.R:228]: You will need to add the correct path for the alarm and telecare extract to the `get_source_extract_path` function.
```suggestion
  ) %>%
  write_rds(get_source_extract_path(year, type = "alarm_telecare", check_mode = "write"))
```

File: Production_scripts/Process_extracts/extract_alarm_telecare.R [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    .data$record_keydate1 >= start_fy(year) &
+      .data$record_keydate2 <= end_fy(year)
+  ) |
+    (
+      .data$record_keydate1 <= end_fy(year) &
+        (
+          .data$record_keydate2 >= start_fy(year) |
+            is_missing(as.character(.data$record_keydate2))
+        )
+    )) %>%
```
Comment (Moohan) [inline Production_scripts/Process_extracts/extract_alarm_telecare.R:200]: Can you use the function `is_date_in_fyyear` here?

File: Production_scripts/Process_extracts/extract_alarm_telecare.R [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+  fs::path("/conf/hscdiip/SLF_Extracts/Social_care")
+
+# Copy and paste from 03-Alarms_Telecare_data.R ----
+## Read Demographic file----
+
+sc_demographics <- haven::read_sav(fs::path(
+  social_care_dir,
+  paste0("sc_demographics_lookup_", latest_update()),
+  ext = "zsav"
+))
```
Comment (Moohan) [inline Production_scripts/Process_extracts/extract_alarm_telecare.R:27]:
```suggestion
sc_demographics <- readr::read_rds(get_sc_demog_lookup_path())
```

File: Production_scripts/Process_extracts/extract_alarm_telecare.R [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+#####################################################
+
+library(createslf)
+library(dplyr)
+library(lubridate)
+
+year <- check_year_format("1920")
+
+social_care_dir <-
+  fs::path("/conf/hscdiip/SLF_Extracts/Social_care")
```
Comment (Moohan) [inline Production_scripts/Process_extracts/extract_alarm_telecare.R:18]: This shouldn't be needed
```suggestion
```

File: Production_scripts/Process_extracts/extract_alarm_telecare.R [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+  distinct(period) %>%
+  mutate(
+    record_date = yq(period) %m+% period(6, "months") %m-% days(1),
+    qtr_start = yq(period) %m+% period(3, "months")
+  )
+
+replaced_start_dates <- at_full_data %>%
+  # Replace missing start dates with the start of the FY
+  left_join(pre_compute_record_dates, by = "period") %>%
+  tidylog::mutate(
```
Comment (Moohan) [inline Production_scripts/Process_extracts/extract_alarm_telecare.R:75]: We can drop the usage of tidylog
```suggestion
  tidylog::mutate(
```

**Review by Moohan — CHANGES_REQUESTED** [review]
A few changes to make. Please request my review again when you've split the code out into separate scripts and had a look at the other changes 👍

**Review by Moohan — CHANGES_REQUESTED** [review]
Looking good, just some minor changes, which I think you should be able to commit directly on GitHub.

I haven't actually tried running the code yet, so I'll do that on my next pass!

File: Production_scripts/Process_extracts/extract_alarm_telecare.R [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+# Description - Match on Client data and Costs data
+#####################################################
+
+library(createslf)
+library(dplyr)
+library(lubridate)
+
+year <- check_year_format("1920")
+
+# Transform B17 ----
```
[the remark this answers is not in the document]
Comment (lizihao-anu) [reply Production_scripts/Process_extracts/extract_alarm_telecare.R:16]: Not from SPSS. Just my personal comment which I forgot to delete

File: Production_scripts/Social_Care/alarm_telecare.R [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+  as_tibble() %>%
+  # Sort for running SPSS
+  arrange(
+    sending_location,
+    social_care_id
+  )
+
+# Save .rds file ----
+qtr_merge %>%
+  write_rds(get_sc_hc_episodes_path(check_mode = "write"))
```
[the remark this answers is not in the document]
Comment (Jennit07) [reply Production_scripts/Social_Care/alarm_telecare.R]: New function `get_sc_at_episodes_path` saves Alarms Telecare data in the same location as other social care extracts

File: Production_scripts/Social_Care/alarm_telecare.R [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+pre_compute_record_dates <- at_full_data %>%
+  distinct(period) %>%
+  mutate(
+    record_date = yq(period) %m+% period(6, "months") %m-% days(1),
+    qtr_start = yq(period) %m+% period(3, "months")
+  )
+
+replaced_start_dates <- at_full_data %>%
+  # Replace missing start dates with the start of the FY
+  left_join(pre_compute_record_dates, by = "period") %>%
```
[the remark this answers is not in the document]
Comment (Jennit07) [reply Production_scripts/Social_Care/alarm_telecare.R]: Changes made

**Comment by Jennit07** [comment]
> Looks good. I would be happy to approve except it looks like the Social Care data wasn't working... I was receiving 0 rows for any query I tried to do!
> 
> Hopefully it will be working next week and @Jennit07 can test the code and give final approval.

I've tested this through and happy it is working as expected. I've committed a few minor changes and it is now ready to merge

**Review by Jennit07 — APPROVED** [review]
I've made a few minor commits, but this is now ready to merge

**Pull request merged**